### PR TITLE
ui: Wire up new KV ember engine to main app

### PR DESCRIPTION
**Pull request by hellobontempo**
A new and improved KV V2 secret engine experience!

This PR switches the default KV view when version=2 to the work implemented in #22426.

Note: this change enforces URL updates for paths within a KV V2 secret engine. For example:
* List view: `/ui/vault/secrets/kv-test/list` -> `/ui/vault/secrets/kv-test/kv/list` (this is automatically redirected)
* List view with subdirectory: `/ui/vault/secrets/kv-test/list/my/` -> `/ui/vault/secrets/kv-test/kv/my%2F/directory`
* Details view: `/ui/vault/secrets/kv-test/show/my/nested-secret` -> `/ui/vault/secrets/kv-test/kv/my%2Fnested-secret/details`

📸 New UI
List view, nested with pagination
Secret details view
Secret metadata view
NEW: Paths tab
Version history

### CODE REVIEW

File: ui/app/models/kmip/scope.js [hunk 1]
```diff
@@ -16,5 +16,5 @@ export default Model.extend({
   attrs: computed(function () {
     return expandAttributeMeta(this, ['name']);
   }),
-  updatePath: lazyCapabilities(apiPath`${'backend'}/scope/${'id'}`, 'id'),
+  updatePath: lazyCapabilities(apiPath`${'backend'}/scope/${'id'}`, 'backend', 'id'),
```
Comment (hashishaw) [inline ui/app/models/kmip/scope.js:19]: Whoops! good catch

File: ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-version-history-diff-test.js
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-version-history-diff-test.js:1]: Thank you for writing this workflow!

File: ui/app/helpers/mountable-secret-engines.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ const ENTERPRISE_SECRET_ENGINES = [
   {
     displayName: 'KMIP',
     type: 'kmip',
-    engineRoute: 'kmip.scopes',
+    engineRoute: 'kmip.scopes.index',
```
Comment (hellobontempo) [inline ui/app/helpers/mountable-secret-engines.js:12]: I had added this, along with a redirect route from the scopes engine and then had subsequently removed it - that commit must have failed.

File: ui/app/models/kmip/scope.js (hunk 1 above)
Comment (hellobontempo) [reply ui/app/models/kmip/scope.js:19]: I should have caught it in the code review! I even thought to check for each arg was present but missed this one 😞 

This was failing a test for me, so I wonder if it was something I changed

File: ui/app/components/mount-backend-form.js [hunk 3]
```diff
@@ -156,7 +157,9 @@ export default class MountBackendForm extends Component {
         this.args.mountType === 'secret' ? 'secrets engine' : 'auth method'
       } at ${path}.`
     );
-    yield this.args.onMountSuccess(type, path, mountModel);
+    // Check whether to use the engine route, since KV version 1 does not
```
Comment (hellobontempo) [inline ui/app/components/mount-backend-form.js:160]: 👏

File: ui/app/helpers/mountable-secret-engines.js (hunk 2 above)
Comment (hellobontempo) [inline ui/app/helpers/mountable-secret-engines.js:12]: IIRC the test failures happening now are from when I tried adding this previously - but I could be wrong!

Comment (hashishaw) [reply ui/app/helpers/mountable-secret-engines.js:12]: 🤔 strange -- maybe we were testing the wrong thing? There doesn't seem to be a `kmip.scopes.js` route though

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 4]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-    await fillIn('[data-test-kv-value]', ''); // clear previous contents
-    await typeIn('[data-test-kv-value]', 'removed!');
-    assert.dom('[data-test-inline-error-message]').doesNotExist('inline error goes away');
-    await click('[data-test-secret-save]');
-    assert
-      .dom('[data-test-message-error]')
-      .includesText(
-        'custom_metadata validation failed: length of key',
-        'shows API error that is not captured by validation'
+  module('mount and configure', function () {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:59]: To make clear which tests are for KV V1 vs V2 I separated into modules. The first module does not set up a secret engine on beforeEach

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 5]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
   hooks.beforeEach(async function () {
     this.uid = uuidv4();
     await authPage.login();
   });
 
-  hooks.afterEach(async function () {
-    this.server.shutdown();
-  });
-
-  test('it creates a secret and redirects', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:80]: Removed -- tested be kv-v2 navigation workflow

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 6]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-      currentRouteName(),
-      'vault.cluster.secrets.backend.show',
-      'redirects to the show page'
-    );
-    await assertSecretWrap(assert, this.server, `${path}/data/${secretPath}`);
-    assert.ok(showPage.editIsPresent, 'shows the edit button');
-    await deleteEngine(path, assert);
-  });
-
-  test('it can create a secret when check-and-set is required', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:110]: moved to v2 module

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 7]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-    assert.strictEqual(
-      currentRouteName(),
-      'vault.cluster.secrets.backend.show',
-      'redirects to the show page'
-    );
-    assert.ok(showPage.editIsPresent, 'shows the edit button');
-    await deleteEngine(enginePath, assert);
-  });
-
-  test('it can create a secret with a non default max version and add metadata', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:127]: Tested on kv-v2 create workflow

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 8]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-    await fillIn('[data-test-kv-value]', 'value');
-    await click('[data-test-save-metadata]');
-    const key = document.querySelector('[data-test-row-label="key"]').innerText;
-    const value = document.querySelector('[data-test-row-value="key"]').innerText;
-    assert.strictEqual(key, 'key', 'metadata key displays after adding it.');
-    assert.strictEqual(value, 'value', 'metadata value displays after adding it.');
-    await deleteEngine(enginePath, assert);
-  });
-
-  skip('it can handle validation on custom metadata', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:165]: Tested on kv-v2 create workflow

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 9]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
+      await consoleComponent.runCommands(`write ${this.backend}/config cas_required=true`);
+      assert.strictEqual(
+        consoleComponent.lastLogOutput,
+        `Success! Data written to: ${this.backend}/config`,
+        'Engine successfully updated'
       );
-    await deleteEngine(enginePath, assert);
-  });
-
-  test('it can mount a KV 2 secret engine with config metadata', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:194]: Moved to mount module

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 10]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-    assert.strictEqual(cas.trim(), 'Yes', 'displays the cas set when configuring the secret-engine');
-    assert.strictEqual(
-      deleteVersionAfter.trim(),
-      '1 second',
-      'displays the delete version after set when configuring the secret-engine'
-    );
-    await deleteEngine(enginePath, assert);
-  });
-
-  test('it can create a secret and metadata can be created and edited', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:235]: removed - tested on kv-v2 create workflow

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 11]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-    const savedMaxVersions = Number(document.querySelectorAll('[data-test-value-div]')[0].innerText);
-    assert.strictEqual(
-      maxVersions,
-      savedMaxVersions,
-      'max_version displays the saved number set when creating the secret'
-    );
-    await deleteEngine(enginePath, assert);
-  });
-
-  test('it shows validation errors', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:262]: removed, tested on kv-v2 create workflow

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 12]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-    await click('[data-test-secret-save]');
-    assert.strictEqual(
-      currentURL(),
-      `/vault/secrets/${enginePath}/show/${secretPath}`,
-      'navigates to show secret'
-    );
-    await deleteEngine(enginePath, assert);
-  });
-
-  test('it navigates to version history and to a specific version', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:298]: kept in v2 module

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 13]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-    assert.dom('[data-test-masked-input]').hasText('bar', 'renders secret on the secret version show page');
-    assert.strictEqual(
-      currentURL(),
-      `/vault/secrets/${enginePath}/show/${secretPath}?version=1`,
-      'redirects to the show page with queryParam version=1'
-    );
-    await deleteEngine(enginePath, assert);
-  });
-
-  test('version 1 performs the correct capabilities lookup and does not show metadata tab', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:332]: Kept all the v1 tests here, in v1 module

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js [hunk 14]
```diff
@@ -66,566 +50,328 @@ const mountEngineGeneratePolicyToken = async (enginePath, secretPath, policy, ve
...
-    await editPage.createSecret(secretPath, 'foo', 'bar');
-    await settled();
-    // use visit helper here because ids with / in them get encoded
-    await visit('/vault/secrets/secret/list/foo/bar');
-    assert.strictEqual(currentRouteName(), 'vault.cluster.secrets.backend.list');
-    assert.ok(currentURL().endsWith('/'), 'redirects to the path ending in a slash');
-    await deleteEngine('secret', assert);
-  });
-
-  test('it can edit via the JSON input', async function (assert) {
```
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:454]: Skipping - getting weird behavior

File: ui/tests/acceptance/secrets/backend/kv/secret-test.js
Comment (hashishaw) [inline ui/tests/acceptance/secrets/backend/kv/secret-test.js:1]: Removed all permissions tests from this file 🎉 (in favor of the kv-v2 workflow tests)

File: ui/app/templates/vault/cluster/secrets/backends.hbs [hunk 15]
```diff
@@ -45,12 +45,13 @@
...
   </ToolbarActions>
 </Toolbar>
+
 {{#each this.sortedDisplayableBackends as |backend|}}
   <LinkedBlock
     @params={{array backend.backendLink backend.id}}
     class="list-item-row linked-block-item is-no-underline"
     data-test-secrets-backend-link={{backend.id}}
-    @disabled={{if backend.isSupportedBackend false true}}
+    @disabled={{not backend.isSupportedBackend}}
```
Comment (kiannaquach) [inline ui/app/templates/vault/cluster/secrets/backends.hbs:54]: 🙃

File: ui/tests/acceptance/enterprise-control-groups-test.js [hunk 16]
```diff
@@ -14,8 +14,7 @@ import authForm from 'vault/tests/pages/components/auth-form';
 import controlGroup from 'vault/tests/pages/components/control-group';
 import controlGroupSuccess from 'vault/tests/pages/components/control-group-success';
 import authPage from 'vault/tests/pages/auth';
-import editPage from 'vault/tests/pages/secrets/backend/kv/edit-secret';
-import listPage from 'vault/tests/pages/secrets/backend/list';
+import { writeSecret } from '../helpers/kv/kv-run-commands';
```
Comment (kiannaquach) [inline ui/tests/acceptance/enterprise-control-groups-test.js:17]: Non-blocking: Can we import `writeSecret` using the direct path instead? `vault/tests/helpers/kv/kv-run-commands`

File: ui/tests/acceptance/secrets/backend/kv/diff-test.js [hunk 17]
```diff
@@ -28,7 +28,7 @@ module('Acceptance | kv2 diff view', function (hooks) {
     this.server.shutdown();
   });
 
-  test('it shows correct diff status based on versions', async function (assert) {
+  test.skip('it shows correct diff status based on versions', async function (assert) {
```
Comment (kiannaquach) [inline ui/tests/acceptance/secrets/backend/kv/diff-test.js:31]: Non-blocking/Nit: 

`import { skip } from 'qunit';`
and use `skip()`

File: ui/app/helpers/mountable-secret-engines.js [hunk 18]
```diff
@@ -129,4 +130,10 @@ export function allEngines() {
   return [...MOUNTABLE_SECRET_ENGINES, ...ENTERPRISE_SECRET_ENGINES];
 }
 
+export function isAddonEngine(type, version) {
+  const engineRoute = allEngines().findBy('type', type)?.engineRoute;
+  if (type === 'kv' && version === 1) return false;
```
Comment (kiannaquach) [inline ui/app/helpers/mountable-secret-engines.js:135]: Nit: Can we do this if statement before the `findBy`?

File: ui/tests/acceptance/secrets/backend/kv/diff-test.js (hunk 17 above)
Comment (hashishaw) [reply ui/tests/acceptance/secrets/backend/kv/diff-test.js:31]: We already import test so I'd rather just use skip from that

**Review by kiannaquach — APPROVED** [review]

**Comment by hashishaw** [comment]
Related follow-on PRs:
#22541 
#22552
#22565
#22593
#22648

**Comment by marcportabellaclotet-mt** [comment]
Does this new version remove the availability to view diffs between secret versions?
With 1.14.3 version I can the diff option, but when using 1.15.0-rcx , the option is no longer in the UI.
I like the new kv UI! great job!

**Comment by hellobontempo** [comment]
We moved the diff to the create a new secret version view (screenshot below 📸 ). We thought comparing version diffs to be more useful when editing. However, if you were using diff comparison across versions, we'd love to know your use case and get feedback on this change. 😁 
Screenshot 2023-09-18 at 2 17 33 PM

**Comment by marcportabellaclotet-mt** [comment]
We are presently employing the 'diff' tool to contrast secret versions across different points in time. Occasionally, we encounter problematic deployments due to incorrect key-value (KV) settings, and it is important for us to pinpoint when this modification was initially introduced.

Displaying the 'diff' at the time of creation is beneficial. Therefore, having both options, one at the moment of creation and the ability to compare any version, would be highly appreciated.

Thank you.

**Comment by hellobontempo** [comment]
Thank you @marcportabellaclotet-mt for your incredibly helpful feedback! PR https://github.com/hashicorp/vault/pull/23200 returns this functionality to the UI and we plan to ship it in the 1.15 GA release. The view will  be located in toolbar of the `Version history` tab. Let me know if you have any further input!
diff

**Comment by marcportabellaclotet-mt** [comment]
Incredible update! We greatly appreciate your attention to this feature.
Vault is outstanding!

**Pull request merged**